state_tracker.go

```diff
@@ -9,13 +9,125 @@ import (
+// run and the resulting state is can be resumed from without incurring data
+// loss. The same `SerializableState` can be used as an input to `Ferry`, which
+// if specified will instruct to `Ferry` to start from a particular location.
+//
+// Code Design
+// -----------
+//
+// In a Ghostferry run, there are two "stages" of operation: the copy stage and
+// the verify stage. Both stages must emit their states to the state tracker in
+// order for them to be interruptible and resumable. These two stages are very
+// similar: they both iterate over the data and tail the binlog. However, there
+// are some minor differences. Example: the verifier stage needs to keep track
+// of the reverify store while the copy stage doesn't.
+//
+// In order to not repeat code, "base structs" are created for the state
+// tracking and serializable state: `BinlogAndIterationStateTracker` and
+// `BinlogAndIterationSerializableState`. `BinlogAndIterationStateTracker` have
+// most of the code/definitions required for the state tracker of both stages
+// to function.  To customize them for the different stages, `CopyStateTracker`
+// and `VerifierStateTracker` contain a minor amount of customized code due to
+// the small differences in requirements. The same can be said about
+// `BinlogAndIterationSerializableState` with respect to
+// `CopySerializableState` and `VerifierSerializableState`.
+//
+// These two stages of state tracker (and serializable states) are owned by a
+// "global" `StateTracker`. This struct keeps two member variables pointing to
+// the state tracker of each stage and that's it. The Ferry creates a new
+// instance of the global `StateTracker` and passes out references of the
+// applicable stage of state tracker to components like `BatchWriter` and
+// `IterativeVerifier`.
+//
+// To summarize, what we have is (arrows point from member variables to owner
+// structs):
+//
+//      BinlogAndIterationStateTracker
+//         |                     |
+//         v                     v
+//  CopyStateTracker    VerifierStateTracker
+//         |                     |
+//         +----------+----------+
+//                    |
+//                    v
+//               StateTracker
+//                    |
+//                    v
+//                  Ferry
+//
+// The same relationship exists for the serializable states.
+
+type BinlogAndIterationSerializableState struct {
+	LastSuccessfulPrimaryKeys map[string]uint64
+	CompletedTables           map[string]bool
+	LastWrittenBinlogPosition mysql.Position
+}
+
+type CopySerializableState struct {
+	*BinlogAndIterationSerializableState
+}
+
+type VerifierSerializableState struct {
+	*BinlogAndIterationSerializableState
+
+	// This is not efficient because we have to build this map of a different
+	// type from the original ReverifyStore struct.
+	//
+	// TODO: address this inefficiency later.
+	ReverifyStore map[string][]uint64
```

Comment:
This is kind of a placeholder for now as I haven't hooked up the VerifierStage to the IterativeVerifier yet. We can iterate on this in the next PR.